### PR TITLE
[FLINK-26519][table] Remove FlinkTypeFactory singleton

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR removes the FlinkTypeFactory singleton

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/spec/DynamicTableSourceSpec.java [hunk 1]
```diff
@@ -66,7 +67,8 @@ public DynamicTableSourceSpec(
         this.sourceAbilities = sourceAbilities;
     }
 
-    private DynamicTableSource getTableSource(FlinkContext flinkContext) {
+    private DynamicTableSource getTableSource(
+            FlinkContext flinkContext, FlinkTypeFactory flinkTypeFactory) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/spec/DynamicTableSourceSpec.java:71]: nit: remove `flink` prefixes, everything is `flink` in this code base

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalHashAggRule.scala [hunk 2]
```diff
@@ -25,6 +25,8 @@ import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalAggregate
 import org.apache.flink.table.planner.plan.nodes.physical.batch.BatchPhysicalHashAggregate
 import org.apache.flink.table.planner.plan.utils.PythonUtil.isPythonAggregate
 import org.apache.flink.table.planner.plan.utils.{AggregateUtil, OperatorType}
+import org.apache.flink.table.planner.utils.ShortcutUtils
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalHashAggRule.scala:28]: all changes in this class are unrelated to the actual issue

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalWindowAggregateRule.scala [hunk 3]
```diff
@@ -27,10 +27,12 @@ import org.apache.flink.table.planner.plan.`trait`.FlinkRelDistribution
 import org.apache.flink.table.planner.plan.logical.{LogicalWindow, SlidingGroupWindow, TumblingGroupWindow}
 import org.apache.flink.table.planner.plan.nodes.FlinkConventions
 import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalWindowAggregate
-import org.apache.flink.table.planner.plan.nodes.physical.batch.{BatchPhysicalLocalSortWindowAggregate, BatchPhysicalSortWindowAggregate, BatchPhysicalHashWindowAggregate, BatchPhysicalLocalHashWindowAggregate}
+import org.apache.flink.table.planner.plan.nodes.physical.batch.{BatchPhysicalHashWindowAggregate, BatchPhysicalLocalHashWindowAggregate, BatchPhysicalLocalSortWindowAggregate, BatchPhysicalSortWindowAggregate}
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalWindowAggregateRule.scala:30]: all changes in this class are unrelated to the actual issue

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/IncrementalAggregateRule.scala [hunk 4]
```diff
@@ -24,6 +24,8 @@ import org.apache.flink.table.planner.calcite.{FlinkContext, FlinkTypeFactory}
 import org.apache.flink.table.planner.plan.PartialFinalType
 import org.apache.flink.table.planner.plan.nodes.physical.stream.{StreamPhysicalExchange, StreamPhysicalGlobalGroupAggregate, StreamPhysicalIncrementalGroupAggregate, StreamPhysicalLocalGroupAggregate}
 import org.apache.flink.table.planner.plan.utils.AggregateUtil
+import org.apache.flink.table.planner.utils.ShortcutUtils
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/IncrementalAggregateRule.scala:27]: all changes in this class are unrelated to the actual issue

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala [hunk 5]
```diff
@@ -712,16 +712,14 @@ object AggregateUtil extends Enumeration {
...
     if (indexOfCountStar.isEmpty) {
-
       val count1 = AggregateCall.create(
         SqlStdOperatorTable.COUNT,
         false,
         false,
         new util.ArrayList[Integer](),
         -1,
-        typeFactory.createFieldTypeFromLogicalType(new BigIntType()),
+        new BasicSqlType(FlinkTypeSystem.INSTANCE, SqlTypeName.BIGINT),
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala:722]: This is not possible. Types need to be `canonized` and need to go through the type factory once in order to allow reference equality.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/CorrelateUtil.scala [hunk 6]
```diff
@@ -59,11 +59,11 @@ object CorrelateUtil {
 
   def projectCorrelateOutputType(
       originalType: RelDataType,
-      projectableFieldSet: Set[Int]): (RelDataType, ListBuffer[Int]) = {
+      projectableFieldSet: Set[Int],
+      flinkTypeFactory: FlinkTypeFactory): (RelDataType, ListBuffer[Int]) = {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/CorrelateUtil.scala:63]: nit: `flinkTypeFactory` -> `typeFactory` here and at all other locations

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RankUtil.scala [hunk 7]
```diff
@@ -24,6 +24,7 @@ import org.apache.flink.table.planner.codegen.ExpressionReducer
 import org.apache.flink.table.planner.plan.nodes.calcite.Rank
 import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalRank
 import org.apache.flink.table.planner.plan.nodes.physical.stream.StreamPhysicalDeduplicate
+import org.apache.flink.table.planner.utils.ShortcutUtils
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RankUtil.scala:27]: unnecessary change

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @slinkydeveloper. Most of the changes look good to me, however, the `AggregateUtil` is the biggest blocker that we need to solve. This is one of the main reasons why this change has not been performed yet.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala (hunk 5 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala:722]: Are you sure? I just tried this on the debugger:

```
new BasicSqlType(FlinkTypeSystem.INSTANCE(), SqlTypeName.BIGINT)
  .equals(flinkTypeFactory.createFieldTypeFromLogicalType(new BigIntType().copy(false))
```

And It's true. 

I also debugged `canonize` and that function for `BasicSqlType` is always no-op, the `canonize` function is doing something only for `ObjectSqlType`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/RowType.java [hunk 8]
```diff
@@ -180,7 +180,7 @@ public int getFieldCount() {
 
     public int getFieldIndex(String fieldName) {
         for (int i = 0; i < fields.size(); i++) {
-            if (fields.get(i).getName().equals(fieldName)) {
+            if (fields.get(i).getName().equalsIgnoreCase(fieldName)) {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/RowType.java:183]: Reverted

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/MetadataTestUtil.scala [hunk 9]
```diff
@@ -272,7 +272,7 @@ object MetadataTestUtil {
 
     val catalogTable = getCatalogTable(resolvedSchema)
 
-    val typeFactory = new FlinkTypeFactory(new FlinkTypeSystem)
+    val typeFactory = new FlinkTypeFactory()
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/MetadataTestUtil.scala:275]: Why not saving it in a static class variable?

Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/MetadataTestUtil.scala:275]: because in the next PR this is going to require `ClassLoader` as parameter

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkTypeSystem.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+import org.apache.calcite.rel.type.RelDataTypeFactoryImpl;
+import org.apache.calcite.rel.type.RelDataTypeSystemImpl;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.flink.table.planner.utils.ShortcutUtils.unwrapTypeFactory;
+
+/** Custom type system for Flink. */
+public class FlinkTypeSystem extends RelDataTypeSystemImpl {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkTypeSystem.java:39]: add `@Internal`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/abilities/source/SourceAbilityContext.java [hunk 11]
```diff
@@ -45,12 +45,16 @@
  * </ul>
  */
 public class SourceAbilityContext implements FlinkContext {
+
     private final RowType sourceRowType;
     private final FlinkContext context;
+    private final FlinkTypeFactory typeFactory;
 
-    public SourceAbilityContext(FlinkContext context, RowType sourceRowType) {
+    public SourceAbilityContext(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/abilities/source/SourceAbilityContext.java:53]: nit: `FlinkContext context, FlinkTypeFactory typeFactory, RowType sourceRowType` it is otherwise weird to have the row type between context parameters

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/MatchCodeGenerator.scala [hunk 12]
```diff
@@ -678,6 +678,7 @@ class MatchCodeGenerator(
         matchAgg.inputExprs.indices.map(i => s"TMP$i"))
 
       val aggInfoList = AggregateUtil.transformToStreamAggregateInfoList(
+        relBuilder.getTypeFactory.asInstanceOf[FlinkTypeFactory],
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/MatchCodeGenerator.scala:681]: you could use `ShortcutUtils` here as well

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnInterval.scala [hunk 13]
```diff
@@ -565,7 +566,10 @@ class FlinkRelMdColumnInterval private extends MetadataHandler[ColumnInterval] {
           inputType: RelDataType,
           isBounded: Boolean): AggregateCall = {
         val outputIndexToAggCallIndexMap = AggregateUtil.getOutputIndexToAggCallIndexMap(
-          aggCalls, inputType, isBounded)
+          ShortcutUtils.unwrapTypeFactory(input),
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnInterval.scala:569]: how about we do a static import for all `ShortcutUtils.unwrapTypeFactory` to make the code more concise. I added a similar feedback to Marios PR recently.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala [hunk 14]
```diff
@@ -404,12 +422,13 @@ object AggregateUtil extends Enumeration {
       indexOfExistingCountStar: Option[Int],
       isStateBackedDataViews: Boolean,
       needDistinctInfo: Boolean,
-      isBounded: Boolean): AggregateInfoList = {
+      isBounded: Boolean) = {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala:425]: @slinkydeveloper still why this?

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala:425]: typo, removed it

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkTypeSystem.java [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+    public RelDataType deriveDecimalPlusType(
+            RelDataTypeFactory typeFactory, RelDataType type1, RelDataType type2) {
+        return deriveDecimalType(
+                typeFactory, type1, type2, LogicalTypeMerging::findAdditionDecimalType);
+    }
+
+    @Override
+    public RelDataType deriveDecimalModType(
+            RelDataTypeFactory typeFactory, RelDataType type1, RelDataType type2) {
+        if (SqlTypeUtil.isExactNumeric(type1)
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkTypeSystem.java]: Reorganized a bit

**Review by twalthr — APPROVED** [review]
Thanks for the update @slinkydeveloper. LGTM.
